man/sslio.8
.TH sslio 8
.SH NAME
sslio \- SSL input/output for service programs
.SH SYNOPSIS
.B sslio
[\-cv]
[\-u
.IR user ]
[\-U
.IR user ]
[\-/
.IR root ]
[\-C
.IR cert ]
[\-K
.IR key ]
[\-A
.IR ca ]
.I prog
.SH DESCRIPTION
.B sslio
provides SSL encrypted network connections for service programs started by
.BR tcpsvd (8)
or
.BR tcpserver (1),
and
.BR tcpclient (1).
.P
Normally
.B sslio
is started by
.BR tcpsvd (8)
or
.BR tcpclient(1),
in turn starts the service program
.IR prog ,
and runs as child process of the service program.
After performing the SSL handshake,
.B sslio
reads SSL encrypted data from the network, and writes decrypted data to the
service program
.IR prog ;
it reads data from the service program
.IR prog ,
and writes SSL encrypted data to the network.
.B sslio
should run under a different user ID than the service program, and with a
changed root directory.
When started by root, the \-u option must be given, and the \-U and \-/
options should be given.
.P
The
.B sslio
program uses the SSLv3 implementation of the matrixssl library.
.SH OPTIONS
.TP
.I prog
.I prog
consists of one or more arguments, specifying the service program normally
run directly by
.BR tcpsvd (8),
or
.BR tcpserver (1).
.TP
.B \-u \fI[:]user[:group]
drop permissions.
Set uid and gid to the
.IR user 's
uid and gid, as found in
.IR /etc/passwd ,
before reading data from, or writing data to the network.
If
.I user
is followed by a colon and a
.IR group ,
set the gid to
.IR group 's
gid, as found in
.IR /etc/group ,
instead of
.IR user 's
gid.
If
.I group
consists of a colon-separated list of group names,
set the group ids of all listed groups.
If
.I user
is prefixed with a colon, the
.I user
and all
.I group
arguments are interpreted as uid and gids respectively, and not looked up in
the password or group file.
All supplementary groups are removed.
This option must be set when
.B sslio
is started by root, and cannot be set otherwise.
.TP
.B \-U \fI[:]user[:group]
drop permissions.
Set uid and gid to the
.IR user 's
uid and gid, as found in
.IR /etc/passwd ,
before running
.IR prog .
If
.I user
is followed by a colon and a
.IR group ,
set the gid to
.IR group 's
gid, as found in
.IR /etc/group ,
instead of
.IR user 's
gid.
If
.I group
consists of a colon-separated list of group names,
set the group ids of all listed groups.
If
.I user
is prefixed with a colon, the
.I user
and all
.I group
arguments are interpreted as uid and gids respectively, and not looked up in
the password or group file.
All supplementary groups are removed.
This option should be set when
.B sslio
is started by root, and cannot be set otherwise.
.TP
.B \-/ \fIroot
chroot.
Change the root directory to
.I root
before reading data from, or writing data to the network.
This option should be set when
.B sslio
is started by root, and cannot be set otherwise.
.TP
.B \-C \fIcert
cert file (server mode).
Read the certificate from the file
.I cert
(default is ``./cert.pem'').
If the \-/ option is given, first the root directory is changed, then the
cert file is read.
.TP
.B \-K \fIkey
private key (server mode).
Read the private key from the file
.I key
(default is
.IR cert ).
If the \-/ option is given, first the root directory is changed, then the
private key is read.
.TP
.B \-A \fIca
ca file (client mode).
Read the trusted root certificate from the file
.IR ca .
Multiple files can be specified, using a semicolon as delimiter.
If the \-/ option is given, first the root directory is changed, then the
ca file is read.
.TP
.B \-c
client mode.
This option must be given when running
.B sslio
under
.BR tcpclient (1).
In client mode, filedescriptors 6 and 7 are used instead of standard input
and standard ouput to read from and write to the network and the service
program.
If the \-A option is given,
.B sslio
refuses to connect to a servers which's certificates cannot be verified by
the root certificates, it accepts any server certificate otherwise.
.TP
.B \-v
verbose.
Print verbose messages to standard error.
.TP
.B \-vv
more verbose.
Print more verbose messages to standard error.
.TP
.B \-vvv
even more verbose.
Print even more verbose messages to standard error.
.SH ENVIRONMENT
.TP
SSLIO_BUFIN
The environment variable
.I SSLIO_BUFIN
overrides the default input buffer size for
.B sslio
(8192).
.TP
SSLIO_BUFOU
The environment variable
.I SSLIO_BUFOU
overrides the default output buffer size for
.B sslio
(12288).
If the output buffer is too small to hold encrypted or decrypted data,
.B sslio
automatically blows up the buffer to
.I SSLIO_BUFOU
more bytes.
.TP
SSLIO_BAD_CERTIFICATE
(client mode)  If the environment variable
.I SSLIO_BAD_CERTIFICATE
is set,
.B sslio \-c
accepts server ceritificates it would normally reject with
.
 fatal: ssl decode error: bad certificate
.TP
SSLIO_HANDSHAKE_TIMOUT
The environment variable
.I SSLIO_HANDSHAKE_TIMEOUT
overrides the default number of seconds
.B sslio
will try to complete the ssl handshake (300).
If the handshake isn't completed after this number of seconds,
.B sslio
exits.
.SH SEE ALSO
sslsvd(8),
tcpsvd(8),
udpsvd(8),
ipsvd(7),
ipsvd-instruct(5),
ipsvd-cdb(8)
.P
http://smarden.org/ipsvd/
.SH AUTHOR
Gerrit Pape <pape@smarden.org>


package/CHANGES
ipsvd 0.14.0
Sat, 19 Jan 2008 00:17:20 +0000
  * uidgid.c, uidgid.h: support numerical uid, gids through :uid:gid...
    (taken from the runit package).
  * tcpsvd.c, udpsvd.c, sslio.c, ssl_io.c: support numerical uid, gids,
    and colon-separated list of groups through :uid:gid... with the -u,
    -U arguments (thx Nikola Vladov).
  * man/tcpsvd.8, man/udpsvd.8, man/sslsvd.8, man/sslio.8, man/ipsvd.7:
    adapt.

ipsvd 0.13.0
Sun, 26 Aug 2007 11:04:21 +0000
  * ipsvd_check.c, man/ipsvd-instruct.5: lines starting with # are
    comments, ignore them completely.
  * ssl_io.c, man/sslio.8: support env variable SSLIO_BAD_CERTIFICATE to
    optionally accept 'bad certificate's (thx Nikola Vladov).
  * ssl_io.c, man/sslio.8, man/sslsvd.8: support env variable
    SSL_HANDSHAKE_TIMEOUT to disconnect a client if the ssl handshake
    isn't completed within some number of seconds (thx Nikola Vladov);
    add ENVIRONMENT section to man/sslsvd.8.
  * Makefile: build with matrixssl-1-8-3-open.

ipsvd 0.12.1
Sat, 04 Feb 2006 19:14:23 +0000
  * Makefile: fix build failure with local matrixssl library (thx Charlie
    Brady).

ipsvd 0.12.0
Fri, 03 Feb 2006 20:26:27 +0000
  * ssl_io.c, ssl_io.h: new; split off most code from sslio.c.
  * sslio.c: main() function only, uses ssl_io() from ssl_io.
  * tcpsvd.c: be sslsvd using ssl_io() #ifdef SSLSVD.
  * sslsvd.c: new; #define SSLSVD, #include tcpsvd.c.
  * man/sslsvd.8: new; man page for sslsvd(8).
  * man/ipsvd-cdb.8, man/ipsvd-instruct.5, man/ipsvd.7, man/sslio.8,
    man/tcpsvd.8, man/udpsvd.8: add SEE ALSO sslsvd(8).
  * sslsvd.check, sslsvd.dist: selftest for sslsvd not yet implemented.
  * doc/index.html: mention sslsvd.

ipsvd 0.11.1
Tue, 18 Oct 2005 19:02:18 +0000
  * sslio.check: tweak sslio program check (fixes testing failure on
    Debian m68k)
  * ipsvd_log.c: minor
  * Makefile, doc/install.html: adapt to matrixssl-1-7-1-open.tar.gz.

ipsvd 0.11.0
Sun, 20 Feb 2005 16:13:38 +0000
  * man/tcpsvd.8: typo (thx Bastian Kleineidam).
  * sslio.check, sslio.dist, sslio.pem: new; ssl program check.
  * Makefile: don't limit sslio program to linux systype; run ssl program
    check; minor
  * package/compile, package/upgrade: adapt.
  * doc/benefits.html, doc/index.html, doc/install.html: sslio works on
    MacOSX; adapt.

ipsvd 0.10.1
Sun, 16 Jan 2005 10:34:19 +0000
  * sslerror_str.c: fix build failure due to broken header include (thx
    Charlie Brady)
  * Makefile: target sslerror_str.o depends on matrixSsl.h.
  * sslio.c: minor.

ipsvd 0.10.0
Sat, 15 Jan 2005 11:47:55 +0000
  * doc/install.html: typo.
  * sslerror_str.c, sslerror_str.h: new; ssl error strings.
  * sslio.c: improve logging, error messages, use sslerror_str; new option
    -c client support to run under tcpclient.
  * uidgid.c: improve error handling.
  * Makefile, man/sslio.8: adapt.
  * package/versions: new.
  * package/COPYING: adapt, 2005.

ipsvd 0.9.7
Fri, 03 Dec 2004 10:15:37 +0000
  * ipsvd_check.c: catch all failures of openreadclose(); disregard
    hostnames (looked up in dns) starting with a dot; fix cdb check on check
    hostname instruction with non-existent forward to match instruction
    directory behavior (deny connection).
  * ipsvd_hostname.c: properly handle failure and nxdomain/noptr of
    dns_name4().
  * doc/examples.html: minor.
  * tcpsvd.c: bug: socka didn't get updated after ECONNABORTED from accept();
    fix: re-set socka_size for each connection (thx Sergei Kolobov).

ipsvd 0.9.6
Mon, 02 Aug 2004 19:18:50 +0000
  * sslio.c: minor cleanup; adapt few verbose log messages; detect
    successful handshake through matrixSslHandshakeIsComplete().
  * doc/install.html, Makefile: use library from src/matrixssl.tar.gz.

ipsvd 0.9.5
Wed, 14 Jul 2004 19:43:48 +0000
  * man/sslio.8: notations.
  * ipsvd_check.c: only parse concurrency instruction if phcc is enabled.
  * tcpsvd.c: bug: concurrency instruction C0 denied connection, fix: C0
    correctly disables phcc (thx Torne).
  * package/compile, package/install-man, package/upgrade: mkdir -p dir ->
    mkdir -p dir/.; ignore grep's return code (fixes build failure on
    solaris, thx Uffe Jakobsen).
  * Makefile: target check-tcpsvd, check-udpsvd: add libraries needed for
    socket() (fixes testing failure on solaris, thx Uffe Jakobsen); use
    matrixssl-1.1.2 with sslio.
  * doc/index.html: minor.

ipsvd 0.9.4
Sat, 19 Jun 2004 14:47:09 +0000
  * doc/benefits.html, man/tcpsvd.8: typos (thx David Thiel).
  * doc/install.html, Makefile: use matrixssl-1.1.1 with sslio.
  * tcpsvd.check, udpsvd.check: use -llocalhost.

ipsvd 0.9.3
Wed, 02 Jun 2004 19:27:13 +0000
  * sslio.c: end network connection when seeing eof from prog; adapt some
    verbose log messages.
  * doc/install.html, Makefile: use matrixssl-1.1 with sslio.

ipsvd 0.9.2
Sat, 15 May 2004 14:11:54 +0000
  * doc/examples.html: use html named entities; minor.
  * man/sslio.8: clarify that chroot is done before keys are read (thx
    Charlie Brady).
  * sslio.c: include <sys/types.h>; ignore sig_pipe, catch sig_term; send
    ssl close alert on end of connection, and fatal errors.
  * use matrixssl-1-0-2.

ipsvd 0.9.1
Mon, 03 May 2004 17:08:41 +0000
  * Makefile, sslio.c: fix build failure due to broken header include (thx
    Charlie Brady).
  * doc/benefits.html: minor.

ipsvd 0.9.0
Sat, 01 May 2004 14:29:26 +0000
  * src/print-cc.sh, src/print-ld.sh: head -1 -> head -n1 (thx Richard A
    Downing).
  * sslio.c, man/sslio.8: new sslio program for Linux.
  * Makefile.matrixssl: new; conditionally build static library.
  * doc/benefits.html: new; some ipsvd's benefits.
  * package/check: run program testing process.
  * doc/examples.html: add sslio bincimaps, sslio fnord.
  * doc/index.html, doc/install.html: adapt.
  * package/compile, package/upgrade: conditionally build, install sslio.
  * package/install: run package/check.

ipsvd 0.8.2
Mon, 23 Feb 2004 20:15:29 +0000
  * man/ipsvd-instruct.5: on C0[:msg] per host concurrency is disabled (thx

src/sslio.c
#include <sys/types.h>
#include <unistd.h>
#include "matrixSsl.h"
#include "uidgid.h"
#include "prot.h"
#include "error.h"
#include "strerr.h"
#include "fd.h"
#include "pathexec.h"
#include "stralloc.h"
#include "byte.h"
#include "taia.h"
#include "iopause.h"
#include "ndelay.h"
#include "fmt.h"
#include "scan.h"
#include "sgetopt.h"
#include "env.h"
#include "sig.h"
#include "sslerror_str.h"
#include "ssl_io.h"

#define USAGEROOT " -u user [-U user] [-/ root] [-C cert] [-K key] [-A ca] [-vc] prog"
#define USAGE " [-C cert] [-K key] [-A ca] [-cv] prog"
#define VERSION "$Id$"

const char *progname;
unsigned int verbose =0;

void usage() {
  if (getuid() == 0) strerr_die4x(111, "usage: ", progname, USAGEROOT, "\n");
  strerr_die4x(111, "usage: ", progname, USAGE, "\n");
}

int main(int argc, const char **argv) {
  int opt;

  progname =*argv;
  pid =getpid();
  id[fmt_ulong(id, pid)] =0;

  while ((opt =getopt(argc, argv, "u:U:/:C:K:A:cvV")) != opteof) {
    switch(opt) {
    case 'u': ssluser =(char*)optarg; break;
    case 'U': svuser =(char*)optarg; break;
    case '/': root =(char*)optarg; break;
    case 'C': cert =(char*)optarg; break;
    case 'K': key =(char*)optarg; break;
    case 'c': client =1; break;
    case 'A': ca =(char*)optarg; break;
    case 'v': ++verbose; break;
    case 'V': strerr_warn1(VERSION, 0);
    case '?': usage();
    }
  }
  argv +=optind;
  if (! argv || ! *argv) usage();

  if (getuid() == 0) { if (! ssluser) usage(); }
  else { if (root || ssluser || svuser) usage(); }

  if (! client) {
    if (! cert) cert ="./cert.pem";
    if (! key) key =cert;
  }
  if (ssluser) if (! uidgids_get(&sslugid, ssluser)) {
    if (errno)
      strerr_die3sys(111, "sslio[", id, "]: fatal: unable to get user/group: ");
    strerr_die4x(100, "sslio[", id, "]: fatal: unknown user/group: ", ssluser);
  }
  if (svuser) if (! uidgids_get(&ugid, svuser)) {
    if (errno)
      strerr_die3sys(111, "sslio[", id, "]: fatal: unable to get user/group: ");
    strerr_die4x(100, "sslio[", id, "]: fatal: unknown user/group: ", svuser);
  }

  return(ssl_io(1, argv));
}
